Sources/Bond/Shared/ViewControllerLifeCycle.swift

```diff
@@ -0,0 +1,126 @@
+//
+//  ViewControllerLifeCycle.swift
+//  Bond-iOS
+//
+//  Created by Ibrahim Koteish on 21/03/2020.
+//  Copyright © 2020 Swift Bond. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+import ReactiveKit
+/// Observe UIViewController life cycle events
+public final class ViewControllerLifeCycle {
+  private let wrapperViewController: WrapperViewController
+  public var viewDidLoad: SafeSignal<Void> {
+    return self.wrapperViewController.viewDidLoadSubject.toSignal()
+  }
+  public var viewWillAppear: SafeSignal<Void> {
+    return self.wrapperViewController.viewWillAppearSubject.toSignal()
+  }
+  public var viewDidAppear: SafeSignal<Void> {
+    return self.wrapperViewController.viewDidAppearSubject.toSignal()
+  }
+  public var viewWillDisappear: SafeSignal<Void> {
+    return self.wrapperViewController.viewWillDisappearSubject.toSignal()
+  }
+  public var viewDidDisappear: SafeSignal<Void> {
+    return self.wrapperViewController.viewDidDisAppearSubject.toSignal()
+  }
+  public init(viewController: UIViewController) {
+    self.wrapperViewController = WrapperViewController()
+    if viewController.isViewLoaded {
+      self.addAsChildViewController(viewController)
+    } else {
+      viewController
+        .reactive
+        .keyPath(\.view, startWithCurrentValue: false)
+        .prefix(maxLength: 1)
+        .observeNext { [weak viewController] _ in
+          guard let viewController = viewController else { return }
+          self.addAsChildViewController(viewController) // weak self ?
+      }
+        .dispose(in: viewController.bag)
+    }
+  }
+
+  private func addAsChildViewController(_ viewController: UIViewController) {
+    viewController.addChild(self.wrapperViewController)
+    viewController.view.addSubview(self.wrapperViewController.view)
+    self.wrapperViewController.view.frame = .zero
+    self.wrapperViewController.view.autoresizingMask = []
+    self.wrapperViewController.didMove(toParent: viewController)
+  }
+}
+
+private extension ViewControllerLifeCycle {
+  final class WrapperViewController: UIViewController {
+    fileprivate let viewDidLoadSubject       = SafeReplayOneSubject<Void>()
+    fileprivate let viewWillAppearSubject    = SafeReplayOneSubject<Void>()
+    fileprivate let viewDidAppearSubject     = SafeReplayOneSubject<Void>()
+    fileprivate let viewWillDisappearSubject = SafeReplayOneSubject<Void>()
+    fileprivate let viewDidDisAppearSubject  = SafeReplayOneSubject<Void>()
```

Comment:
Instead of having a subject for each event, how about we do something like:

```swift
    public enum LifecycleEvent {
        case viewDidLoad
        case viewWillAppear
        case viewDidAppear
        case viewWillDisappear
        case viewDidDisappear
    }

    deinit {
        _lifecycleEvent.send(completion: .finished)
    }

    private let _lifecycleEvent = PassthroughSubject<LifecycleEvent, Never>()

    public var lifecycleEvents: Signal<LifecycleEvent, Never> {
        var signal = _lifecycleEvent.toSignal()
        if isViewLoaded {
            signal = signal.prepend(.viewDidLoad)
        }
        return signal
    }

    public func lifecycleEvent(_ event: LifecycleEvent) -> Signal<Void, Never> {
        return lifecycleEvents.filter { $0 == event }.eraseType()
    }
```